File: service/internal/users/usecases.go
package users

import (
	"context"
	"errors"
)

var (
	// ErrUserNotFound indicates that no user was found with the provided ID.
	ErrUserNotFound = errors.New("users: The requested user was not found")
)

// GetUserUseCase is the use case used to load a single user by ID.
type GetUserUseCase interface {
	// GetUserByID allows us to load the user with the given ID.
	GetUserByID(ctx context.Context, id UserID) (UserModel, error)
}

File: service/src/users/usecases.rs
use super::{UserID, UserModel};
use async_trait::async_trait;

#[async_trait]
pub trait GetUserUseCase {
    /// Get the user that has the given User ID.
    ///
    /// # Parameters
    /// - `user_id` - The ID of the user to get
    ///
    /// # Returns
    /// The user, if it was found. `None` if it doesn't exist.
    async fn get_user_by_id(&self, user_id: &UserID) -> Option<UserModel>;
}
